microsetta_admin/server.py

```diff
@@ -774,6 +775,217 @@ def scan():
                                       recorded_type)
 
 
+def _get_color_code():
+    val = '0123456789ABCDEF'
+    code = ["#"+''.join([random.choice(val) for i in range(6)])]
+    return code[0]
+
+
+def _get_legends(criteria):
+
+    dict = {}
+    if criteria not in session:
+        status, fields = APIRequest.get(
+            '/api/admin/barcode_query_fields')
+
+        if status == 200:
+            for obj in fields:
+                if criteria == "Type" and obj["label"] == "Sample Site":
+                    for val in obj["values"]:
+                        color_code = _get_color_code()
+                        while color_code in dict:
+                            color_code = _get_color_code()
+                        dict[obj["values"][val]] = color_code
+
+                if criteria == "Status" and obj["label"] == "Sample Status":
+                    for val in obj["values"]:
+                        color_code = _get_color_code()
+                        while color_code in dict:
+                            color_code = _get_color_code()
+                        tmp = obj["values"][val].replace(' ', '-').lower()
+                        dict[tmp] = color_code
+
+                if criteria == "Project" and obj["label"] == "Project":
+                    for val in obj["values"]:
+                        if not ('-' in obj["values"][val]):
+                            color_code = _get_color_code()
+                            while color_code in dict:
+                                color_code = _get_color_code()
+                            dict[obj["values"][val]] = color_code
+
```

Comment:
Should the function return an error if it's called with an invalid criteria?

Reply:
In case the function is called with an invalid criteria, it will return an empty dictionary.

Reply:
It looks like there's nothing checking that between the function and the UI. Should there be a check and explicit exception if the result is an empty dictionary, rather than letting the UI silently fail?

Reply:
In case of invalid criteria, visualization must show all samples to "None" legend for provided criteria. I have updated code to return the legend dictionary to at least return "None" as value for every sorting criteria.